### PR TITLE
Use CMAKE_CXX_STANDARD to set C++ standard

### PR DESCRIPTION
**Pull request by amadio**
Fixes: [ROOT-9695](https://sft.its.cern.ch/jira/browse/ROOT-9695), [ROOT-9805](https://sft.its.cern.ch/jira/browse/ROOT-9805).

### CODE REVIEW

File: cmake/modules/CheckCompiler.cmake [hunk 1]
```diff
@@ -97,43 +97,29 @@ endif()
...
-endif()
-if(cxx14)
-  CHECK_CXX_COMPILER_FLAG("-std=c++14" HAS_CXX14)
-  if(NOT HAS_CXX14)
-    message(STATUS "Current compiler does not suppport -std=c++14 option. Switching OFF cxx14 option")
-    set(cxx14 OFF CACHE BOOL "" FORCE)
-  endif()
+#---C++ standard----------------------------------------------------------------------
+
+set(CMAKE_CXX_STANDARD 11)
```
Comment (amadio) [inline cmake/modules/CheckCompiler.cmake]: Note: setting `CMAKE_CXX_STANDARD` on the command line will override this.

**Comment by amadio** [comment]
@phsft-bot build

**Comment by amadio** [comment]
Turns out `try_compile()` ignores `CMAKE_CXX_STANDARD` for CMake <= 3.8... How nice.

**Comment by amadio** [comment]
I'm reverting the commits that set the version based on Git just to be able to build the PR.

File: cmake/modules/CheckCompiler.cmake [hunk 2]
```diff
@@ -97,43 +97,31 @@ endif()
...
-    set(root7 OFF CACHE BOOL "" FORCE)
-  endif()
-    set(http ON CACHE BOOL "" FORCE)
+
+set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
+set(CMAKE_CXX_EXTENSIONS FALSE)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++${CMAKE_CXX_STANDARD}")
+
+if(root7 AND (NOT http OR CMAKE_CXX_STANDARD EQUAL 11))
```
Comment (Axel-Naumann) [inline cmake/modules/CheckCompiler.cmake]: That does not look equivalent to the original logic. I.e. `-Droot7=On -Dcxx11=On` would trigger the warning, whatever the setting of `http`, and with `root7` enabled, `http` would be `On`.

File: cmake/modules/RootConfiguration.cmake [hunk 3]
```diff
@@ -493,31 +493,25 @@ if(veccore)
...
   set(hasveccore undef)
 endif()
-if(cxx11)
-  set(cxxversion c++11)
-  set(usec++11 define)
-else()
+
+set(cxxversion c++${CMAKE_CXX_STANDARD})
+
+if(CMAKE_CXX_STANDARD EQUAL 17)
```
Comment (Axel-Naumann) [inline cmake/modules/RootConfiguration.cmake]: Can we `undef` all, and then `define` `usec++${CMAKE_CXX_STANDARD}`?

File: cmake/modules/SearchInstalledSoftware.cmake [hunk 4]
```diff
@@ -1507,12 +1507,12 @@ endif()
...
 
-  if(cxx11)
+  if(CMAKE_CXX_STANDARD EQUAL 11)
     find_package(CUDA 7.5)
-  elseif(cxx14)
+  elseif(CMAKE_CXX_STANDARD EQUAL 14)
     message(STATUS "Detected request for c++14, requiring minimum version CUDA 9.0 (default 7.5)")
     find_package(CUDA 9.0)
-  elseif(cxx17)
+  elseif(CMAKE_CXX_STANDARD EQUAL 17)
```
Comment (Axel-Naumann) [inline cmake/modules/SearchInstalledSoftware.cmake]: What will happen for 20? Why not just `else()`?

File: core/base/CMakeLists.txt [hunk 5]
```diff
@@ -6,7 +6,7 @@ ROOT_ADD_TEST_SUBDIRECTORY(test)
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/res ${CMAKE_CURRENT_SOURCE_DIR}/../foundation/res)
 
-if(cxx14 OR cxx17 OR root7)
+if(CMAKE_CXX_STANDARD GREATER 11)
```
Comment (Axel-Naumann) [inline core/base/CMakeLists.txt]: :+1:

File: interpreter/CMakeLists.txt [hunk 6]
```diff
@@ -52,9 +52,10 @@ set(LLVM_EXTERNAL_PROJECTS "cling" CACHE STRING "")
 # to serialize access to llvm.  We can later review how to make this finer grained by using llvm's own locking
 # mechanism.
 set(LLVM_ENABLE_THREADS OFF CACHE BOOL "")
-if(cxx17)
+
+if(CMAKE_CXX_STANDARD EQUAL 17)
   set(LLVM_ENABLE_CXX1Z ON CACHE BOOL "" FORCE)
-elseif(cxx14)
+elseif(CMAKE_CXX_STANDARD EQUAL 14)
   set(LLVM_ENABLE_CXX1Y ON CACHE BOOL "" FORCE)
```
Comment (Axel-Naumann) [inline interpreter/CMakeLists.txt]: That looks surprising to me - why don't we turn on `LLVM_ENABLE_CXX14`? And yes, that should probably not be part of this PR.

File: cmake/modules/CheckCompiler.cmake (hunk 2 above)
Comment (amadio) [reply cmake/modules/CheckCompiler.cmake]: I can make it more or less equivalent to before, like this:
```cmake
if(root7)
  if(CMAKE_CXX_STANDARD EQUAL 11)
    message(WARNING "ROOT 7 requires C++14 or higher. Switching OFF root7 option.")
  else()
    set(http ON CACHE BOOL "" FORCE)
  endif()
endif()
```
Note that `http` is turned ON regardless in the current code.

File: cmake/modules/CheckCompiler.cmake [hunk 7]
```diff
@@ -97,43 +97,31 @@ endif()
...
 endif()
-if(root7)
-  if(cxx11)
-    message(STATUS "ROOT7 interfaces require >= cxx14 which is disabled. Switching OFF root7 option")
-    set(root7 OFF CACHE BOOL "" FORCE)
-  endif()
-    set(http ON CACHE BOOL "" FORCE)
+
+set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
+set(CMAKE_CXX_EXTENSIONS FALSE)
```
Comment (Axel-Naumann) [inline cmake/modules/CheckCompiler.cmake]: Please leave them on if people asked for `On`.

File: cmake/modules/RootConfiguration.cmake (hunk 3 above)
Comment (amadio) [reply cmake/modules/RootConfiguration.cmake]: Yes, good idea. I will do it like that.

File: cmake/modules/SearchInstalledSoftware.cmake (hunk 4 above)
Comment (amadio) [reply cmake/modules/SearchInstalledSoftware.cmake]: We need support in other places like `root-config`, `R__HAS_CXX20`, etc, before we can let people use C++20, that's why I added the message. We can fix that before 6.18 is out, though, but it will require more work. We also need to make sure nothing that has been removed is still used by ROOT, and so on.

File: interpreter/CMakeLists.txt (hunk 6 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: I don't know, I just thought that `LLVM_ENABLE_CXX1Y` is the only way to tell it to enable C++14...

File: cmake/modules/SearchInstalledSoftware.cmake (hunk 4 above)
Comment (Axel-Naumann) [reply cmake/modules/SearchInstalledSoftware.cmake]: Ideally, ROOT (not caring about > 17 but only about >= 17) should set `R__HAS_CXX17` when compiled with C++20. That shows a design mistake in these `usec++...`: we should always define the earlier ones, too, such that for `CMAKE_CXX_STANDARD` >= 17, all current ones are defined.

This should probably go into a separate PR...

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

File: cmake/modules/SearchInstalledSoftware.cmake (hunk 4 above)
Comment (Axel-Naumann) [reply cmake/modules/SearchInstalledSoftware.cmake]: We agreed to rip out `R__USE_CXX` from ROOT; it was ill-defined (e.g. `R__USE_CXX14` was not defined for C++17) and people should not rely on it, but use `__cplusplus` instead. Which is what we'll do in the single use in ROOT. @amadio offered a subsequent PR.

**Comment by amadio** [comment]
Fixed switching C++ standard also with current (deprecated) options. Also, old options prevail if both are set, e.g. using `-Dcxx11=ON and -DCMAKE_CXX_STANDARD=14` will turn on C++11. By default `CMAKE_CXX_STANDARD` is set to `11` and can be overwritten by the command line. Changing standards after the initial configuration should always work now.

File: cmake/modules/SearchInstalledSoftware.cmake (hunk 4 above)
Comment (amadio) [reply cmake/modules/SearchInstalledSoftware.cmake]: Actually, it was easy enough that I just added this to the current PR (removal of `R__USE_CXX*`).

File: interpreter/CMakeLists.txt (hunk 6 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: Just for future reference:
```
epsft-53 root $ git grep LLVM_ENABLE_CXX
interpreter/CMakeLists.txt:  set(LLVM_ENABLE_CXX1Z ON CACHE BOOL "" FORCE)
interpreter/CMakeLists.txt:  set(LLVM_ENABLE_CXX1Y ON CACHE BOOL "" FORCE)
interpreter/llvm/src/CMakeLists.txt:option(LLVM_ENABLE_CXX1Y "Compile with C++1y enabled." OFF)
interpreter/llvm/src/CMakeLists.txt:option(LLVM_ENABLE_CXX1Z "Compile with C++1z enabled." OFF)
interpreter/llvm/src/cmake/modules/HandleLLVMOptions.cmake:  if (LLVM_ENABLE_CXX1Y)
interpreter/llvm/src/cmake/modules/HandleLLVMOptions.cmake:  elseif(LLVM_ENABLE_CXX1Z)
interpreter/llvm/src/docs/CMake.rst:**LLVM_ENABLE_CXX1Y**:BOOL
```

File: cmake/modules/RootConfiguration.cmake [hunk 8]
```diff
@@ -574,6 +550,8 @@ else()
   set(hastmvagpu undef)
 endif()
 
+unset(found_stdstringview CACHE)
```
Comment (Axel-Naumann) [inline cmake/modules/RootConfiguration.cmake]: Add comment: `// Allow change of CMAKE_CXX_STANDARD`?

File: cmake/modules/SearchInstalledSoftware.cmake (hunk 4 above)
Comment (Axel-Naumann) [inline cmake/modules/SearchInstalledSoftware.cmake]: Can we now make this `else()`?

File: cmake/modules/CheckCompiler.cmake [hunk 9]
```diff
@@ -97,43 +97,42 @@ endif()
...
-  endif()
-endif()
-if(cxx14)
-  CHECK_CXX_COMPILER_FLAG("-std=c++14" HAS_CXX14)
-  if(NOT HAS_CXX14)
-    message(STATUS "Current compiler does not suppport -std=c++14 option. Switching OFF cxx14 option")
-    set(cxx14 OFF CACHE BOOL "" FORCE)
+#---C++ standard----------------------------------------------------------------------
+
+set(CMAKE_CXX_STANDARD 11 CACHE STRING "")
```
Comment (Axel-Naumann) [inline cmake/modules/CheckCompiler.cmake]: That seems surprising. Doesn't that override a user setting? Can we not simply take the compiler default as our default?

File: cmake/modules/CheckCompiler.cmake [hunk 10]
```diff
@@ -97,43 +97,42 @@ endif()
...
+set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
+set(CMAKE_CXX_EXTENSIONS FALSE CACHE BOOL "")
```
Comment (Axel-Naumann) [inline cmake/modules/CheckCompiler.cmake]: Is this not overriding the user setting anymore?

File: cmake/modules/RootConfiguration.cmake (hunk 8 above)
Comment (amadio) [reply cmake/modules/RootConfiguration.cmake]: Sure, will do.

File: cmake/modules/SearchInstalledSoftware.cmake (hunk 4 above)
Comment (amadio) [reply cmake/modules/SearchInstalledSoftware.cmake]: Yes, maybe. This changes depending on the version of CUDA and the host compiler. Maybe it's best to use a `try_compile()` in the future to check if things actually work and error out if not.

File: cmake/modules/CheckCompiler.cmake (hunk 9 above)
Comment (amadio) [reply cmake/modules/CheckCompiler.cmake]: No, the command line will override this, since this is setting in the cache. It only overrides if `FORCE` is used.

File: cmake/modules/CheckCompiler.cmake (hunk 10 above)
Comment (amadio) [reply cmake/modules/CheckCompiler.cmake]: Overrides only with `FORCE` at the end. See `cmake --help-command set`.

File: cmake/modules/RootConfiguration.cmake (hunk 8 above)
Comment (amadio) [reply cmake/modules/RootConfiguration.cmake]: Done.

File: cmake/modules/SearchInstalledSoftware.cmake (hunk 4 above)
Comment (pcanal) [reply cmake/modules/SearchInstalledSoftware.cmake]: yes, we should really on try_compile for most things .. as long as they get rerun/recalculated for every relevant changes (for example when the user change the requested C++ standard)

File: config/root-config.in [hunk 11]
```diff
@@ -146,7 +145,7 @@ fi
 ### machine dependent settings ###
 
 # Set the C++ standard version
-cxxversionflag="-std=${cxxversion} "
+cxxversionflag="-std=c++@CMAKE_CXX_STANDARD@ "
```
Comment (pcanal) [inline config/root-config.in]: Is that really correct?  (i.e. Isn't CMAKE_CXX_STANDARD=17 requires -std=c++1z on some compiler?)

Comment (pcanal) [reply config/root-config.in]: Okay ... are we saying that we will not support any compiler/standard pair that is not fully numerical? (i.e. so no 20 support until the compiler officially use --std=c++20)?  [That is a valid choice that ought to be 'documented' somewhere and imply that we do not support experimenting with upcoming standard, only established ones ... or am I missing something]

Comment (amadio) [reply config/root-config.in]: No, but that's the status quo. We only support C++11, 14, and 17. Support for C++20 will require more work than just adding the flag in root-config, as was the case for C++17 due to removed things from older C++ standards, etc.

**Comment by amadio** [comment]
@phsft-bot build

**Comment by Axel-Naumann** [comment]
@pcanal do we need to keep `R__USE_CXX11` and friends? Could be as easy as `#if __cplusplus > ... #define R__USE_CXX11`. I'd still prefer to get rid of them as there is no usage in ROOT.

**Comment by pcanal** [comment]
> do we need to keep R__USE_CXX11 and friends? Could be as easy as #if __cplusplus > ... #define R__USE_CXX11. I'd still prefer to get rid of them as there is no usage in ROOT.

Yes.  The R__USE were brought to life at a point of time where __cplusplus was not well defined *and* was not a good proxy for the compiler's actual feature set ... and we had not yet thought of and/or decide to purse try_compile (or configure equivalent).

Nowadays, I think we should almost never really on __cplusplus but rather on try_compile (and thus furthermore not on R__USE_CXX11)
